pkg/agent/proxy/proxier.go

```diff
@@ -797,6 +805,68 @@ func (p *proxier) OnServiceSynced() {
 	}
 }
 
+// OnNodeAdd is called whenever creation of new node object
+// is observed.
+func (p *proxier) OnNodeAdd(node *corev1.Node) {
+	if node.Name != p.hostname {
+		return
+	}
+
+	if reflect.DeepEqual(p.nodeLabels, node.Labels) {
+		return
+	}
+
+	p.serviceEndpointsMapsMutex.Lock()
+	p.nodeLabels = map[string]string{}
+	for k, v := range node.Labels {
+		p.nodeLabels[k] = v
+	}
+	p.serviceEndpointsMapsMutex.Unlock()
+	klog.V(4).InfoS("Updated proxier node labels", "labels", node.Labels)
+
+	p.syncProxyRules()
+}
+
+// OnNodeUpdate is called whenever modification of an existing
+// node object is observed.
+func (p *proxier) OnNodeUpdate(oldNode, node *corev1.Node) {
+	if node.Name != p.hostname {
+		return
+	}
+
+	if reflect.DeepEqual(p.nodeLabels, node.Labels) {
+		return
+	}
+
+	p.serviceEndpointsMapsMutex.Lock()
+	p.nodeLabels = map[string]string{}
+	for k, v := range node.Labels {
+		p.nodeLabels[k] = v
+	}
+	p.serviceEndpointsMapsMutex.Unlock()
+	klog.V(4).InfoS("Updated proxier node labels", "labels", node.Labels)
+
+	p.syncProxyRules()
+}
+
+// OnNodeDelete is called whenever deletion of an existing node
+// object is observed.
+func (p *proxier) OnNodeDelete(node *corev1.Node) {
+	if node.Name != p.hostname {
```

Comment:
question: seems the proxier is processing Node Add/Delete event for itself.  how can this happen?

Reply:
Copied from kube-proxy

Reply:
https://github.com/kubernetes/kubernetes/blob/02462739ca544dc61f4bf00a290d247c453030ef/pkg/proxy/iptables/proxier.go#L675 @xliuxu